src/main/java/org/mockito/internal/stubbing/answers/ThrowsException.java

```diff
@@ -14,27 +14,35 @@
 import static org.mockito.internal.exceptions.Reporter.cannotStubWithNullThrowable;
 import static org.mockito.internal.exceptions.Reporter.checkedExceptionInvalid;
 
+/**
+ * An answer that always throws the same throwable.
+ */
 public class ThrowsException implements Answer<Object>, ValidableAnswer, Serializable {
 
     private static final long serialVersionUID = 1128820328555183980L;
     private final Throwable throwable;
     private final ConditionalStackTraceFilter filter = new ConditionalStackTraceFilter();
 
+    /**
+     * Creates a new answer always throwing the given throwable. If it is null,
+     * {@linkplain ValidableAnswer#validateFor(InvocationOnMock) answer validation}
+     * will fail.
+     */
     public ThrowsException(Throwable throwable) {
         this.throwable = throwable;
     }
 
     public Object answer(InvocationOnMock invocation) throws Throwable {
-        if (MockUtil.isMock(throwable)) {
-            throw throwable;
+        if (throwable == null) {
+            throw new IllegalStateException("throwable is null: " +
```

Comment:
This exception message is not consistent with how we handle exceptions. Take a look at a Reporter class. Typically we throw an exception that is or derives from MockitoException. The message needs to be very clean, actionable to the end user. I suggest to create a high level unit test that demonstrates this use case from the standpoint of the end user. Run this test, observe the exception, would a newbie Mockito understand the exception message and was able to resolve the problem quickly?

Reply:
This one is tricky, as it must not occur because of an _external_ precondition that _Mockito_ first validates this answer. If this precondition is met (i.e., Mockito never calls `answer` if `validateFor` fails), it will never fire. Therefore, I can't think of how to yield this exception through _public_ API.

I've checked on 2.x with the following test, `answer` throws an NPE:

```java
    @Test
    public void should_not_answer_null_throwable() throws Throwable {
        try {
            // Shall #validateFor first!
            new ThrowsException(null).answer(new InvocationBuilder().toInvocation());
            Assertions.fail("Should have raised NPE");
        } catch (NullPointerException expected) {}
    }
```

I see such alternatives:
1. Keep an implicit null check. I believe it's not OK because this check is performed on the object state, and not on a user-supplied argument.
2. Add an explicit null check, throw NPE with description.
3. Add an explicit null check, throw IllegalStateException with description.
4. Add an explicit null check, report in a more user-friendly format.
5. Add a `validateFor` check in `answer`. It will _definetely_ catch when validateFor isn't called, but is redundant in case when Mockito always validates this answer (= adheres to the Answer+ValidableAnswer contract).

I'd say (5) is the safest one, but you make the same check twice (externally in the framework code + internally in the class itself). I can't say how expensive that is, or shall we trust the rest of the framework here, what do you think?

It seems `Reporter` is used to supply user-facing exceptions, but this one shall never occur in correct Mockito implementation. How do you report _internal_ errors?

Reply:
Oh, you're right, the user will never see it when using the public API. I missed that.

If this is only internal, IllegalStateException is perfect.